libraries/standard/mqtt/utest/mqtt_lightweight_utest.c

```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
+    willInfo.payloadLength = ( uint16_t ) strlen( willInfo.pPayload );
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    ( void ) memset( &willInfo, 0x00, sizeof( MQTTPublishInfo_t ) );
+    willInfo.retain = true;
+    willInfo.qos = MQTTQoS1;
+    willInfo.pTopicName = "test";
+    willInfo.topicNameLength = ( uint16_t ) strlen( willInfo.pTopicName );
+    willInfo.pPayload = "test";
+    willInfo.payloadLength = ( uint16_t ) strlen( willInfo.pPayload );
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+
+    /* Again with QoS 2 and 0. */
+
+    willInfo.qos = MQTTQoS2;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+
+    willInfo.qos = MQTTQoS0;
+    willInfo.retain = false;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+
+    /* For this example, MQTT_GetConnectPacketSize() will return
+     * packetSize = remainingLength +2 (two byte fixed header).
+     * Make sure MQTT_SerializeConnect()
+     * fails  when remaining length is more than packet size. */
+    /* TODO: Since MQTT_SerializeConnect trusts the remaining length to be
+     * accurate, either change the implementation or remove this test, since it
+     * should fail. */
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength + 4, &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Tests that MQTT_GetSubscribePacketSize works as intended.
+ */
+void test_MQTT_GetSubscribePacketSize( void )
+{
+    MQTTSubscribeInfo_t subscriptionList;
+    size_t subscriptionCount = 0;
+    size_t remainingLength = 0;
+    size_t packetSize = 0;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTSubscribeInfo_t fourThousandSubscriptions[ 4096 ] = { 0 };
```

Comment:
This test will increase the required stack size by at least 4KB * sizeof(MQTTSubscribeInfo_t ).  As long as unit tests are not targeted to run on targets that are memory constrained, this should be ok.